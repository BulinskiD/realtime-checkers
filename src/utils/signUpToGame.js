import { firestore } from "../api/firebase";
import handleError from "./handleError";

export default async (email, gameID, players) => {
  const color = players.length === 0 ? "white" : "black";
  players.push({ email, started: false, color });
  try {
    await firestore
      .collection("games")
      .doc(gameID)
      .update({
        players: players
      });
    await firestore
      .collection("users")
      .doc(email)
      .update({
        gameID: gameID
      });
  } catch (error) {
    handleError(error);
  }
};
