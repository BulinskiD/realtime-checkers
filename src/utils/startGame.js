import seedCheckers from "./seedCheckers";
import {firestore} from "../api/firebase";

export default async (gameState) => {

    const board = seedCheckers();

    const {selectedChecker, activePoles, ...withoutChecker} = gameState;

    try {
        await firestore.collection("games").doc(gameState.id).set({
            ...withoutChecker,
            status: 'white',
            checkersPosition: board
        });
    } catch (error) {
        console.log(error);
    }
}