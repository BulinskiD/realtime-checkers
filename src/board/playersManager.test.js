import React from 'react';
import configureMockStore from 'redux-mock-store';
import ConnectedPlayersManager, {PlayersManager} from './playersManager';
import { shallow } from 'enzyme';
import startGame from '../utils/startGame';
import signUpToGame from '../utils/signUpToGame';


jest.mock('../utils/startGame');
jest.mock('../utils/signUpToGame');
let data;
beforeEach(()=> {
    const currentGame = {currentGame: {status: 'not-started', players: [], selectedChecker: {col: 2, row: 2},
        checkersPosition: [{col: 2, row: 2, color: 'black', selected: false}], nextMove: false}};
    data = {...currentGame, user: "test", canActivateGame: false, gameAvailable: true};
});

describe("PlayersManager", () => {
    it('Should display sign_up button for given props', () => {
          const component = shallow(<PlayersManager {...data} />);
          expect(component.find('Button').text()).toBe("Dołącz do gry");
          expect(component).toMatchSnapshot();
    });

    it('Should invoke sign_up_to_game function on button clicked', () => {
        const component = shallow(<PlayersManager {...data} gameID={'22'} />);
        component.find('Button').simulate('click');
        expect(signUpToGame).toHaveBeenCalledWith('test', '22', data.currentGame.players);
    });

    it('Should display start_game button for given props', () => {
        data.canActivateGame = true;
        data.gameAvailable = false;
        data.currentGame.status = 'not-started';
        const component = shallow(<PlayersManager {...data} />);
        expect(component.find('Button').text()).toBe("Zacznij grę");
        expect(component).toMatchSnapshot();
    });

    it('Should invoke start_game function on button clicked', () => {
        data.canActivateGame = true;
        data.gameAvailable = false;
        data.currentGame.status = 'not-started';
        const component = shallow(<PlayersManager {...data} />);
        component.find('Button').simulate('click');
        expect(startGame).toHaveBeenCalledWith(data.currentGame, "test");
    });
});

describe("mapStateToProps", () => {
    const store = configureMockStore();
   it('Should map state to props correctly with isParticipant == false && gameAvailable === true', () => {
       data.currentGame.players = [];
       const mockedStore = store({...data});
       const component = shallow(<ConnectedPlayersManager store={mockedStore} gameID={'22'} />);
       expect(component.find('PlayersManager').props().currentGame).toStrictEqual({...data.currentGame});
       expect(component.find('PlayersManager').props().canActivateGame).toBe(false);
       expect(component.find('PlayersManager').props().gameAvailable).toBe(true);
   });

    it('Should map state to props correctly with isParticipant == false && gameAvailable === false', () => {
        data.currentGame.players = ['test', 'test'];
        const mockedStore = store({...data});
        const component = shallow(<ConnectedPlayersManager store={mockedStore} gameID={'22'} />);
        expect(component.find('PlayersManager').props().currentGame).toStrictEqual({...data.currentGame});
        expect(component.find('PlayersManager').props().canActivateGame).toBe(false);
        expect(component.find('PlayersManager').props().gameAvailable).toBe(false);
    });

    it('Should map state to props correctly with isParticipant == false && gameAvailable === false', () => {
        data.currentGame.players = [{email: 'test', started: false}, {email: 'test2', started: false}];
        data.user = {email: 'test'};
        const mockedStore = store({...data});
        const component = shallow(<ConnectedPlayersManager store={mockedStore} gameID={'22'} />);
        expect(component.find('PlayersManager').props().currentGame).toStrictEqual({...data.currentGame});
        expect(component.find('PlayersManager').props().canActivateGame).toBe(true);
        expect(component.find('PlayersManager').props().gameAvailable).toBe(false);
    });
});