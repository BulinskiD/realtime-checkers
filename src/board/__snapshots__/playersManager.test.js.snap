// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayersManager Should display list of players 1`] = `
<styled.div>
  <GameInfo
    status="not-started"
  />
  <Button
    active={false}
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Dołącz do gry
  </Button>
  <strong>
    Aktywni gracze
  </strong>
  <styled.div>
    <styled.span
      key="test"
    >
      test
    </styled.span>
    <styled.span
      key="test2"
    >
      test2
    </styled.span>
  </styled.div>
</styled.div>
`;

exports[`PlayersManager Should display sign_up button for given props 1`] = `
<styled.div>
  <GameInfo
    status="not-started"
  />
  <Button
    active={false}
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Dołącz do gry
  </Button>
  <strong>
    Aktywni gracze
  </strong>
  <styled.div />
</styled.div>
`;

exports[`PlayersManager Should display start_game button for given props 1`] = `
<styled.div>
  <GameInfo
    status="not-started"
  />
  <Button
    active={false}
    className="start-game"
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Zacznij grę
  </Button>
  <strong>
    Aktywni gracze
  </strong>
  <styled.div />
</styled.div>
`;
